Uploader/TraceFileScanner.cs

```diff
@@ -0,0 +1,147 @@
+﻿using NLog;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Abstractions;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+/// <summary>
+/// Scans the trace directory for trace files that are ready to upload or archive.
+/// </summary>
+public class TraceFileScanner
+{
+    private static readonly Logger logger = LogManager.GetCurrentClassLogger();
+    private static readonly Regex TraceFileRegex = new Regex(@"^coverage_\d*_\d*.txt$");
+
+    private readonly string traceDirectory;
+    private readonly Regex versionAssemblyRegex;
+    private readonly IFileSystem fileSystem;
+
+    public TraceFileScanner(string traceDirectory, string versionAssembly, IFileSystem fileSystem)
+    {
+        this.traceDirectory = traceDirectory;
+        this.versionAssemblyRegex = new Regex(@"^Assembly=" + Regex.Escape(versionAssembly) + @".*Version:([^ ]*).*", RegexOptions.IgnoreCase);
+        this.fileSystem = fileSystem;
+    }
+
+    /// <summary>
+    /// Returns all trace files that can be uploaded or archived.
+    /// </summary>
+    public IEnumerable<ScannedFile> ListTraceFilesReadyForUpload()
+    {
+        List<string> files;
+        try
+        {
+            files = fileSystem.Directory.EnumerateFiles(traceDirectory).ToList();
+        }
+        catch (Exception e)
+        {
+            logger.Error(e, "Unable to list files in {traceDirectory}. Will retry later", traceDirectory);
+            yield break;
+        }
+        
+        logger.Debug("Scanning {fileCount} files", files.Count);
+        foreach (string filePath in files)
+        {
+            string fileName = Path.GetFileName(filePath);
+            if (!IsTraceFile(fileName))
+            {
+                logger.Debug("Skipping file that does not look like a trace file: {unknownFilePath}", filePath);
+                continue;
+            }
+
+            ScannedFile scannedFile = ScanFile(filePath);
+            if (scannedFile != null)
+            {
+                yield return scannedFile;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Scans the given file path and returns the resulting ScannedFile or null in case the file should be ignored.
+    /// </summary>
+    private ScannedFile ScanFile(string filePath)
```

Comment:
put this in constructor of scanned file or static construction method?

Reply:
then I have to pass both the logger and the IFileSystem to that new method. I find that less readable than the current version TBH. In my mind ScannedFile is a pure immutable data class (although I didn't know how to make it properly immutable so it's acutally mutable :-( )